src/sentry/integrations/bitbucket/integration.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import absolute_import
+
+from sentry.integrations import IntegrationProvider, IntegrationMetadata
+from sentry.pipeline import NestedPipelineView
+from sentry.identity.pipeline import IdentityProviderPipeline
+from django.utils.translation import ugettext_lazy as _
+from sentry.utils.http import absolute_uri
+from sentry.models import Integration
+
+DESCRIPTION = """
+Bitbucket for Sentry.io
+"""
+metadata = IntegrationMetadata(
+    description=DESCRIPTION.strip(),
+    author='The Sentry Team',
+    noun=_('Bitbucket Account'),
+    issue_url='https://github.com/getsentry/sentry/issues/new?title=Bitbucket%20Integration:%20&labels=Component%3A%20Integrations',
+    source_url='https://github.com/getsentry/sentry/tree/master/src/sentry/integrations/bitbucket',
+    aspects={},
+)
+
+
+class BitbucketIntegrationProvider(IntegrationProvider):
+    key = 'bitbucket'
+    name = 'Bitbucket'
+    metadata = metadata
+
+    def get_pipeline_views(self):
+        identity_pipeline_config = {
+            'redirect_url': absolute_uri('/extensions/bitbucket/setup/'),
+        }
+        identity_pipeline_view = NestedPipelineView(
+            bind_key='identity',
+            provider_key='bitbucket',
+            pipeline_cls=IdentityProviderPipeline,
+            config=identity_pipeline_config,
+        )
+
+        return [
+            identity_pipeline_view,
+        ]
+
+    def build_integration(self, state):
+        # TODO(LB): Add verification for clientKey
+        if state.get('publicKey'):
+            principal_data = state['principal']
+            user_data = state['user']
+            return {
+                'provider': 'bitbucket',
+                'external_id': state['clientKey'],
+                'name': principal_data['username'],
+                'metadata': {
+                    'public_key': state['publicKey'],
+                    'shared_secret': state['sharedSecret'],
+                    'base_url': state['baseApiUrl'],
+                    'domain_name': principal_data['links']['html']['href'].replace('https://', ''),
+                    'icon': principal_data['links']['avatar']['href'],
+                },
+                'user_identity': {
+                    'type': 'bitbucket',
+                    'name': user_data['username'],
+                    'display_name': user_data['display_name'],
+                    'account_id': user_data['account_id'],
+                    'icon': user_data['links']['avatar']['href'],
+                }
+            }
+
+        integration = Integration.objects.get(
+            provider='bitbucket',
+            external_id=state['identity']['bitbucket_client_id']
+        )
+        return {
```

Comment:
Instead of doing this I think what I would do is refactor the `finish_pipeline` method over in the `integrations.pipeline` module to support being told 'The integration was already created out of band, instead lookup the integration using this given `external_id` and attach it there.

So maybe create a named tuple like `ExistingIntegrationRef(exeternal_id= state['identity']['bitbucket_client_id'])` and that can be returned and in the `finish_pipeline` Instead of get_or_create we just do a `get` and attach it to the organization.

```py
# The integration was installed by the webhook, allow the pipeline to attach it to the organization.
if not state.get('publicKey'):
    return ExistingIntegrationRef(exeternal_id= state['identity']['bitbucket_client_id'])

principal_data = state['principal']
user_data = state['user']
return {
    'provider': 'bitbucket',
    'external_id': state['clientKey'],
    'name': principal_data['username'],
    'metadata': {
        'public_key': state['publicKey'],
        'shared_secret': state['sharedSecret'],
        'base_url': state['baseApiUrl'],
        'domain_name': principal_data['links']['html']['href'].replace('https://', ''),
        'icon': principal_data['links']['avatar']['href'],
    },
    'user_identity': {
        'type': 'bitbucket',
        'name': user_data['username'],
        'display_name': user_data['display_name'],
        'account_id': user_data['account_id'],
        'icon': user_data['links']['avatar']['href'],
    }
}
```

Reply:
@EvanPurkhiser I can add a TODO and follow up with another PR as this is refactoring `IntegrationPipeline` as opposed to being specifically about Bitbucket authentication. I don't want to alter the pipeline in the same PR in case we need to revert it later on (i.e. Bitbucket people tell us something that makes us look for a different way to connect with it).

Reply:
Yeah I feel you. You could even follow up beforehand with it if you want. Up to you, a TODO is totally fine.